streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -630,57 +631,185 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
+                  .flipCauseOption(c)
+                  .map(ZChannel.refailCause(_))
+                  .getOrElse(ZChannel.fromZIO(outDoneSignal.await)),
+              out2 => ZChannel.write(out2) *> readerCh
             )
+          z0
+        }
+
+      val resCh: ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = ZChannel
+        .scoped[Env1](enqueuer)
+        .concatMapWith { enqueueFib =>
+          readerCh
+        }((_, o) => o, (o, _) => o)
+        .embedInput(input)
+
+      resCh
+    }
+
+    ZChannel.unwrap(z)
+  }
+
+  final def mapOutZIOParUnordered[Env1 <: Env, OutErr1 >: OutErr, OutElem2](n: Int, bufferSize: Int = 16)(
+    f: OutElem => ZIO[Env1, OutErr1, OutElem2]
+  )(implicit trace: Trace): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = {
+    val z0: ZIO[Any, Nothing, ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone]] = for {
+      input      <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
+      queueReader = ZChannel.fromInput(input)
+      q          <- zio.Queue.bounded[Any](bufferSize)
+      permits    <- zio.Semaphore.make(n)
+    } yield {
+      def enqueue(a: OutElem): ZIO[Env1, Nothing, Unit] = ZIO.uninterruptibleMask { restore =>
+        for {
+          localScope <- zio.Scope.make
+          _          <- restore(permits.withPermitScoped.provideEnvironment(ZEnvironment(localScope)))
+          fib <- {
+            restore {
+              val z1 = f(a)
+              z1
+                .foldCauseZIO(
+                  c => q.offer(QRes.failCause(c)),
+                  a2 => q.offer(a2)
+                )
+            }
+              .onExit(localScope.close(_))
+              .unit
+              .fork
           }
+        } yield ()
+      }
 
-        consumer.embedInput(input)
+      val foreachCh: ZChannel[Env1, OutErr, OutElem, OutDone, OutErr1, Nothing, OutDone] = {
+        lazy val proc: ZChannel[Env1, OutErr, OutElem, OutDone, OutErr1, Nothing, OutDone] =
+          ZChannel.readWithCause(
+            in => {
+              ZChannel.fromZIO(enqueue(in)) *> proc
+            },
+            ZChannel.refailCause(_),
+            ZChannel.succeedNow(_)
+          )
+
+        proc
+      }
+
+      val enqueuer = queueReader
+        .pipeTo(self)
+        .pipeTo(foreachCh)
+        .runScoped
+        .foldCauseZIO(
+          //this message terminates processing so it's ok for it to race with in-flight computations
+          c => {
+            q.offer(QRes.failCause(c))
+          },
+          done => {
+            //make sure this is the last message in the queue
+            permits
+              .withPermits(n) {
+                q.offer(QRes(done))
+              }
+          }
+        )
+        .forkScoped
+
+      val reader: ZChannel[Any, Any, Any, Any, OutErr1, OutElem2, OutDone] = {
+        lazy val reader0: ZChannel[Any, Any, Any, Any, OutErr1, OutElem2, OutDone] = ZChannel
+          .fromZIO(q.take)
```

Comment:
thought about it, I think it complicates the code a bit...
but I decided not to do so since it violates the buffer size constraint and may effectively lead to double the amount of buffered elements.